### PR TITLE
Add get_data_by_element function to VolumeData

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Allows retrieving data by each element from the H5 file.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/IO/H5/VolumeData.cpp [hunk 1]
```diff
@@ -481,6 +482,120 @@ std::pair<size_t, size_t> offset_and_length_for_grid(
...
+    }
+  }
+  result.shrink_to_fit();
+  // Sort by observation_value
+  alg::sort(result, [](const auto& lhs, const auto& rhs) {
+    return std::get<1>(lhs) < std::get<1>(rhs);
+  });
+
+  // Retrieve element data and insert into result
+  // for (size_t i = 0; i < result.size(); ++i) {
```
Comment (kidder) [inline src/IO/H5/VolumeData.cpp]: remove

File: src/IO/H5/VolumeData.hpp [hunk 2]
```diff
@@ -136,6 +136,20 @@ class VolumeData : public h5::Object {
   /// `observation_id`
   std::vector<std::vector<size_t>> get_extents(size_t observation_id) const;
 
+  /// Retrieve volume data for IDs in
+  /// \f$[start_observation_value, end_observation_value]\f$.
```
Comment (kidder) [inline src/IO/H5/VolumeData.hpp]: this does not render correctly as doxygen treats the `_` as subscript markers

File: src/IO/H5/VolumeData.hpp [hunk 3]
```diff
@@ -136,6 +136,20 @@ class VolumeData : public h5::Object {
...
+  ///
+  /// - If `start_observation_value` is not `std::nullopt` then return
+  ///   everything from the beginning of the data to `end_observation_value`. If
+  ///   `end_observation_value` is `std::nullopt` then return everything from
+  ///   `start_observation_value` to the end of the data.
+  auto get_data_by_element(std::optional<double> start_observation_value,
```
Comment (kidder) [inline src/IO/H5/VolumeData.hpp:153]: please document what is returned (the return type is not showing up in the detailed description).  Also nice to know what the tuple types correspond to.   Also note how everything is sorted.

File: tests/Unit/Helpers/IO/VolumeData.cpp [hunk 4]
```diff
@@ -154,6 +154,48 @@ void check_volume_data(
...
+                           return tensor_component.name == component;
+                         });
+        REQUIRE(component_data_it != volume_data_it->tensor_components.end());
+        DataVector expected_component_data{expected_data.size()};
+        for (size_t k = 0; k < expected_component_data.size(); ++k) {
+          expected_component_data[k] = expected_data[k];
+        }
+        CHECK(std::get<DataVector>(component_data_it->data) ==
+              expected_component_data);
+      }
```
Comment (kidder) [inline tests/Unit/Helpers/IO/VolumeData.cpp:196]: is the sorting checked anywhere?  if not please check it

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
Squashed and rebased. Thanks for the review :)

**Review by kidder — DISMISSED** [review]
LGTM except for the build error

**Review by kidder — APPROVED** [review]

**Pull request merged**